gatsby-source-graphql fails

<!--
  Please fill out each section below, otherwise, your issue will be closed. This info allows Gatsby maintainers to diagnose (and fix!) your issue as quickly as possible.

  Useful Links:
  - Documentation: https://www.gatsbyjs.org/docs/
  - How to File an Issue: https://www.gatsbyjs.org/contributing/how-to-file-an-issue/

  Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
-->

## Description

The latest version of `gatsby-source-graphql` fails because it's missing a file. Having dug into this a bit, I think it's because that file (batching/dataloader-link) is gitignored - https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-graphql/.gitignore#L2 

### Steps to reproduce

Install the latest version
Try building the site

### Expected result

Gatsby should build

### Actual result

Gatsby fails with

```
Error in "[HIDDENFORPRIVACY]/node_modules/gatsby-source-graphql/gatsby-node.js": Cannot find module './batching/dataloader-link'
```

### Environment

```

  System:
    OS: macOS 10.15.3
    CPU: (16) x64 Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 10.16.0 - ~/.nvm/versions/node/v10.16.0/bin/node
    Yarn: 1.19.2 - /usr/local/bin/yarn
    npm: 6.11.3 - ~/.nvm/versions/node/v10.16.0/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Chrome: 80.0.3987.149
    Firefox: 74.0
    Safari: 13.0.5
  npmPackages:
    gatsby: ^2.13.41 => 2.20.12 
    gatsby-image: ^2.1.2 => 2.3.1 
    gatsby-plugin-canonical-urls: ^2.0.13 => 2.2.1 
    gatsby-plugin-emotion: ^4.0.7 => 4.2.1 
    gatsby-plugin-google-tagmanager: ^2.0.15 => 2.2.1 
    gatsby-plugin-manifest: ^2.2.5 => 2.3.3 
    gatsby-plugin-netlify: ^2.1.31 => 2.2.1 
    gatsby-plugin-offline: ^3.0.6 => 3.1.2 
    gatsby-plugin-react-helmet: ^3.0.12 => 3.2.1 
    gatsby-plugin-react-svg: ^2.1.1 => 2.1.2 
    gatsby-plugin-robots-txt: ^1.4.0 => 1.5.0 
    gatsby-plugin-sharp: ^2.1.3 => 2.5.3 
    gatsby-plugin-sitemap: ^2.1.0 => 2.3.1 
    gatsby-plugin-web-font-loader: ^1.0.4 => 1.0.4 
    gatsby-source-filesystem: ^2.0.37 => 2.2.2 
    gatsby-source-graphql: ^2.3.1 => 2.3.1 
    gatsby-source-instagram: ^0.5.0 => 0.5.1 
    gatsby-source-twitter: ^3.0.0 => 3.2.1 
    gatsby-transformer-csv: ^2.2.1 => 2.2.1 
    gatsby-transformer-sharp: ^2.1.21 => 2.4.3 
```

I have the following problem, but I don't know if is related because I'm using .tsx files. This is the code from console:

`$ gatsby develop
success open and validate gatsby-configs - 0.024s

 ERROR

Error in "C:\Users\Calebe\Desktop\www\g-port\node_modules\gatsby-source-graphql\gatsby-node.js": Cannot find module './batching/dataloader-link'
Require stack:
- C:\Users\Calebe\Desktop\www\g-port\node_modules\gatsby-source-graphql\gatsby-node.js
- C:\Users\Calebe\Desktop\www\g-port\node_modules\gatsby\dist\bootstrap\resolve-module-exports.js
- C:\Users\Calebe\Desktop\www\g-port\node_modules\gatsby\dist\bootstrap\load-plugins\validate.js
- C:\Users\Calebe\Desktop\www\g-port\node_modules\gatsby\dist\bootstrap\load-plugins\load.js
- C:\Users\Calebe\Desktop\www\g-port\node_modules\gatsby\dist\bootstrap\load-plugins\index.js
- C:\Users\Calebe\Desktop\www\g-port\node_modules\gatsby\dist\bootstrap\index.js
- C:\Users\Calebe\Desktop\www\g-port\node_modules\gatsby\dist\commands\develop.js
- C:\Users\Calebe\AppData\Roaming\npm\node_modules\gatsby-cli\lib\create-cli.js
- C:\Users\Calebe\AppData\Roaming\npm\node_modules\gatsby-cli\lib\index.js



  Error: Cannot find module './batching/dataloader-link'
  Require stack:
  - C:\Users\Calebe\Desktop\www\g-port\node_modules\gatsby-source-graphql\gatsby  -node.js
  - C:\Users\Calebe\Desktop\www\g-port\node_modules\gatsby\dist\bootstrap\resolv  e-module-exports.js
  - C:\Users\Calebe\Desktop\www\g-port\node_modules\gatsby\dist\bootstrap\load-p  lugins\validate.js
  - C:\Users\Calebe\Desktop\www\g-port\node_modules\gatsby\dist\bootstrap\load-p  lugins\load.js
  - C:\Users\Calebe\Desktop\www\g-port\node_modules\gatsby\dist\bootstrap\load-p  lugins\index.js
  - C:\Users\Calebe\Desktop\www\g-port\node_modules\gatsby\dist\bootstrap\index.  js
  - C:\Users\Calebe\Desktop\www\g-port\node_modules\gatsby\dist\commands\develop  .js
  - C:\Users\Calebe\AppData\Roaming\npm\node_modules\gatsby-cli\lib\create-cli.j  s
  - C:\Users\Calebe\AppData\Roaming\npm\node_modules\gatsby-cli\lib\index.js

  - loader.js:982 Function.Module._resolveFilename
    internal/modules/cjs/loader.js:982:15

  - loader.js:864 Function.Module._load
    internal/modules/cjs/loader.js:864:27

  - loader.js:1044 Module.require
    internal/modules/cjs/loader.js:1044:19

  - v8-compile-cache.js:159 require
    [g-port]/[v8-compile-cache]/v8-compile-cache.js:159:20

  - gatsby-node.js:26 Object.<anonymous>
    [g-port]/[gatsby-source-graphql]/gatsby-node.js:26:5

  - v8-compile-cache.js:178 Module._compile
    [g-port]/[v8-compile-cache]/v8-compile-cache.js:178:30

  - loader.js:1178 Object.Module._extensions..js
    internal/modules/cjs/loader.js:1178:10

  - loader.js:1002 Module.load
    internal/modules/cjs/loader.js:1002:32

  - loader.js:901 Function.Module._load
    internal/modules/cjs/loader.js:901:14

  - loader.js:1044 Module.require
    internal/modules/cjs/loader.js:1044:19

  - v8-compile-cache.js:159 require
    [g-port]/[v8-compile-cache]/v8-compile-cache.js:159:20

  - resolve-module-exports.js:126 module.exports
    [g-port]/[gatsby]/dist/bootstrap/resolve-module-exports.js:126:26

  - validate.js:120
    [g-port]/[gatsby]/dist/bootstrap/load-plugins/validate.js:120:31

  - Array.forEach

  - validate.js:113 collatePluginAPIs
    [g-port]/[gatsby]/dist/bootstrap/load-plugins/validate.js:113:20

  - index.js:62 module.exports
    [g-port]/[gatsby]/dist/bootstrap/load-plugins/index.js:62:13


not finished load plugins - 0.799s
`

My Docker build also fails with error 
```
Error in "/var/www/node_modules/gatsby-source-graphql/gatsby-node.js": Cannot find module './batching/dataloader-link'
Error: Cannot find module './batching/dataloader-link'
```
I tried different versions of gatsby-source-graphql including the latest

I'm getting errors too: `Error in "/Users/glauber/Dev/delivering-hope/node_modules/gatsby-source-graphql/gatsby-node.js": Cannot find module './batching/dataloader-link'`

--- a/packages/gatsby-source-graphql/src/batching/__tests__/dataloader-link.js
+++ b/packages/gatsby-source-graphql/src/batching/__tests__/dataloader-link.js
@@ -1,14 +1,13 @@
-import { parse } from "graphql"
-import { execute } from "apollo-link"
-import { createDataloaderLink } from "../dataloader-link"
+const { parse } = require(`graphql`)
+const { execute } = require(`apollo-link`)
+const { createDataloaderLink } = require(`../dataloader-link`)
 
 const sampleQuery = parse(`{ foo }`)
 const expectedSampleQueryResult = { data: { foo: `bar` } }
 
-// eslint-disable-next-line @typescript-eslint/camelcase
 const fetchResult = { data: { gatsby0_foo: `bar` } }
 
-const makeFetch = (expectedResult: any = fetchResult): jest.Mock<any> =>
+const makeFetch = (expectedResult = fetchResult) =>
   jest.fn(() =>
     Promise.resolve({
       json: () => Promise.resolve(expectedResult),
@@ -23,7 +22,7 @@ describe(`createDataloaderLink`, () => {
     })
     const observable = execute(link, { query: sampleQuery })
     observable.subscribe({
-      next: (result: any) => {
+      next: result => {
         expect(result).toEqual(expectedSampleQueryResult)
         done()
       },

--- a/packages/gatsby-source-graphql/src/batching/__tests__/merge-queries.js
+++ b/packages/gatsby-source-graphql/src/batching/__tests__/merge-queries.js
@@ -1,5 +1,5 @@
-import { print, parse } from "graphql"
-import { IQuery, merge, resolveResult } from "../merge-queries"
+const { print, parse } = require(`graphql`)
+const { merge, resolveResult } = require(`../merge-queries`)
 
 describe(`Query merging`, () => {
   it(`merges simple queries`, () => {
@@ -213,7 +213,7 @@ describe(`Resolving merged query results`, () => {
   })
 
   it(`throws on unexpected results`, () => {
-    const shouldThrow = (): void => {
+    const shouldThrow = () => {
       resolveResult({
         data: {
           gatsby0_foo: `foo`,
@@ -225,10 +225,8 @@ describe(`Resolving merged query results`, () => {
   })
 })
 
-type QueryFixture = [string, object]
-
-function fromFixtures(fixtures: QueryFixture[]): IQuery[] {
-  return fixtures.map(([query, variables]: QueryFixture) => {
+function fromFixtures(fixtures) {
+  return fixtures.map(([query, variables]) => {
     return {
       query: parse(query),
       variables,

--- a/packages/gatsby-source-graphql/src/batching/dataloader-link.js
+++ b/packages/gatsby-source-graphql/src/batching/dataloader-link.js
@@ -1,22 +1,14 @@
-import DataLoader from "dataloader"
-import { ApolloLink, Observable, Operation, FetchResult } from "apollo-link"
-import { print } from "graphql"
-import { IQuery, IQueryResult, merge, resolveResult } from "./merge-queries"
+const DataLoader = require(`dataloader`)
+const { ApolloLink, Observable } = require(`apollo-link`)
+const { print } = require(`graphql`)
+const { merge, resolveResult } = require(`./merge-queries`)
 
-interface IOptions {
-  uri: string
-  fetch: Function
-  fetchOptions?: object
-  dataLoaderOptions?: object
-  headers?: object
-}
-
-export function createDataloaderLink(options: IOptions): ApolloLink {
-  const load = async (keys: ReadonlyArray<IQuery>): Promise<IQueryResult[]> => {
+export function createDataloaderLink(options) {
+  const load = async keys => {
     const query = merge(keys)
-    const result: object = await request(query, options)
+    const result = await request(query, options)
     if (!isValidGraphQLResult(result)) {
-      const error: any = new Error(
+      const error = new Error(
         `Failed to load query batch:\n${formatErrors(result)}`
       )
       error.name = `GraphQLError`
@@ -34,12 +26,12 @@ export function createDataloaderLink(options: IOptions): ApolloLink {
   const dataloader = new DataLoader(load, {
     cache: false,
     maxBatchSize,
-    batchScheduleFn: (callback): any => setTimeout(callback, 50),
+    batchScheduleFn: callback => setTimeout(callback, 50),
     ...options.dataLoaderOptions,
   })
 
   return new ApolloLink(
-    (operation: Operation): Observable<FetchResult> =>
+    operation =>
       new Observable(observer => {
         const { query, variables } = operation
 
@@ -61,7 +53,7 @@ export function createDataloaderLink(options: IOptions): ApolloLink {
   )
 }
 
-function formatErrors(result: any): string {
+function formatErrors(result) {
   if (result?.errors?.length > 0) {
     return result.errors
       .map(error => {
@@ -75,7 +67,7 @@ function formatErrors(result: any): string {
   return `Unexpected GraphQL result`
 }
 
-function isValidGraphQLResult(response): response is IQueryResult {
+function isValidGraphQLResult(response) {
   return (
     response &&
     response.data &&
@@ -83,7 +75,7 @@ function isValidGraphQLResult(response): response is IQueryResult {
   )
 }
 
-async function request(query: IQuery, options: IOptions): Promise<object> {
+async function request(query, options) {
   const { uri, headers = {}, fetch, fetchOptions } = options
 
   const body = JSON.stringify({

--- a/packages/gatsby-source-graphql/src/batching/merge-queries.js
+++ b/packages/gatsby-source-graphql/src/batching/merge-queries.js
@@ -1,35 +1,9 @@
-import {
-  visit,
-  visitInParallel,
-  Kind,
-  DocumentNode,
-  VariableNode,
-  SelectionNode,
-  FragmentSpreadNode,
-  FragmentDefinitionNode,
-  InlineFragmentNode,
-  FieldNode,
-  NameNode,
-  OperationDefinitionNode,
-  Visitor,
-  ASTKindToNode,
-  VariableDefinitionNode,
-  DirectiveNode,
-} from "graphql"
-import _ from "lodash"
-
-export interface IQuery {
-  query: DocumentNode
-  variables: object
-}
-
-export interface IQueryResult {
-  data: object
-}
+const { visit, visitInParallel, Kind } = require(`graphql`)
+const _ = require(`lodash`)
 
 const Prefix = {
-  create: (index: number): string => `gatsby${index}_`,
-  parseKey: (prefixedKey: string): { index: number; originalKey: string } => {
+  create: index => `gatsby${index}_`,
+  parseKey: prefixedKey => {
     const match = /^gatsby([\d]+)_(.*)$/.exec(prefixedKey)
     if (!match || match.length !== 3 || isNaN(Number(match[1])) || !match[2]) {
       throw new Error(`Unexpected data key: ${prefixedKey}`)
@@ -74,13 +48,13 @@ const Prefix = {
  *   }
  *   fragment FooQuery on Query { baz }
  */
-export function merge(queries: ReadonlyArray<IQuery>): IQuery {
-  const mergedVariables: object = {}
-  const mergedVariableDefinitions: VariableDefinitionNode[] = []
-  const mergedSelections: SelectionNode[] = []
-  const mergedFragmentMap: Map<string, FragmentDefinitionNode> = new Map()
+export function merge(queries) {
+  const mergedVariables = {}
+  const mergedVariableDefinitions = []
+  const mergedSelections = []
+  const mergedFragmentMap = new Map()
 
-  queries.forEach((query: IQuery, index: number) => {
+  queries.forEach((query, index) => {
     const prefixedQuery = prefixQueryParts(Prefix.create(index), query)
 
     prefixedQuery.query.definitions.forEach(def => {
@@ -99,7 +73,7 @@ export function merge(queries: ReadonlyArray<IQuery>): IQuery {
     Object.assign(mergedVariables, prefixedQuery.variables)
   })
 
-  const mergedQueryDefinition: OperationDefinitionNode = {
+  const mergedQueryDefinition = {
     kind: Kind.OPERATION_DEFINITION,
     operation: `query`,
     variableDefinitions: mergedVariableDefinitions,
@@ -121,72 +95,60 @@ export function merge(queries: ReadonlyArray<IQuery>): IQuery {
 /**
  * Split and transform result of the query produced by the `merge` function
  */
-export function resolveResult(mergedQueryResult: IQueryResult): IQueryResult[] {
+export function resolveResult(mergedQueryResult) {
   const data = mergedQueryResult.data
 
-  return Object.keys(data).reduce(
-    (acc: IQueryResult[], prefixedKey: string): IQueryResult[] => {
-      const { index, originalKey } = Prefix.parseKey(prefixedKey)
-      if (!acc[index]) acc[index] = { data: {} }
-      acc[index].data[originalKey] = data[prefixedKey]
-      return acc
-    },
-    []
-  )
+  return Object.keys(data).reduce((acc, prefixedKey) => {
+    const { index, originalKey } = Prefix.parseKey(prefixedKey)
+    if (!acc[index]) acc[index] = { data: {} }
+    acc[index].data[originalKey] = data[prefixedKey]
+    return acc
+  }, [])
 }
 
 const Visitors = {
-  detectFragmentsWithVariables: (
-    fragmentsWithVariables: Set<string>
-  ): Visitor<ASTKindToNode> => {
+  detectFragmentsWithVariables: fragmentsWithVariables => {
     let currentFragmentName
     return {
       [Kind.FRAGMENT_DEFINITION]: {
-        enter: (def: FragmentDefinitionNode): void => {
+        enter: def => {
           currentFragmentName = def.name.value
         },
-        leave: (): void => {
+        leave: () => {
           currentFragmentName = null
         },
       },
-      [Kind.VARIABLE]: (): void => {
+      [Kind.VARIABLE]: () => {
         if (currentFragmentName) {
           fragmentsWithVariables.add(currentFragmentName)
         }
       },
     }
   },
-  prefixVariables: (prefix: string): Visitor<ASTKindToNode> => {
+  prefixVariables: prefix => {
     return {
-      [Kind.VARIABLE]: (variable: VariableNode): VariableNode =>
-        prefixNodeName(variable, prefix),
+      [Kind.VARIABLE]: variable => prefixNodeName(variable, prefix),
     }
   },
-  prefixFragmentNames: (
-    prefix: string,
-    fragmentNames: Set<string>
-  ): Visitor<ASTKindToNode> => {
+  prefixFragmentNames: (prefix, fragmentNames) => {
     return {
-      [Kind.FRAGMENT_DEFINITION]: (
-        def: FragmentDefinitionNode
-      ): FragmentDefinitionNode | void =>
+      [Kind.FRAGMENT_DEFINITION]: def =>
         fragmentNames.has(def.name.value) ? prefixNodeName(def, prefix) : def,
-
-      [Kind.FRAGMENT_SPREAD]: (def: FragmentSpreadNode): FragmentSpreadNode =>
+      [Kind.FRAGMENT_SPREAD]: def =>
         fragmentNames.has(def.name.value) ? prefixNodeName(def, prefix) : def,
     }
   },
 }
 
-function prefixQueryParts(prefix: string, query: IQuery): IQuery {
-  let document: DocumentNode = aliasTopLevelFields(prefix, query.query)
+function prefixQueryParts(prefix, query) {
+  let document = aliasTopLevelFields(prefix, query.query)
   const variableNames = Object.keys(query.variables)
 
   if (variableNames.length === 0) {
     return { ...query, query: document }
   }
 
-  const fragmentsWithVariables: Set<string> = new Set()
+  const fragmentsWithVariables = new Set()
 
   document = visit(
     document,
@@ -211,7 +173,7 @@ function prefixQueryParts(prefix: string, query: IQuery): IQuery {
         [Kind.INLINE_FRAGMENT]: [`selectionSet`],
         [Kind.FIELD]: [`selectionSet`],
         [Kind.SELECTION_SET]: [`selections`],
-      } as any
+      }
     )
   }
 
@@ -231,9 +193,9 @@ function prefixQueryParts(prefix: string, query: IQuery): IQuery {
  *
  * @see aliasFieldsInSelection for implementation details
  */
-function aliasTopLevelFields(prefix: string, doc: DocumentNode): DocumentNode {
+function aliasTopLevelFields(prefix, doc) {
   const transformer = {
-    [Kind.OPERATION_DEFINITION]: (def): OperationDefinitionNode => {
+    [Kind.OPERATION_DEFINITION]: def => {
       const { selections } = def.selectionSet
       return {
         ...def,
@@ -244,7 +206,7 @@ function aliasTopLevelFields(prefix: string, doc: DocumentNode): DocumentNode {
       }
     },
   }
-  return visit(doc, transformer, { [Kind.DOCUMENT]: [`definitions`] } as any)
+  return visit(doc, transformer, { [Kind.DOCUMENT]: [`definitions`] })
 }
 
 /**
@@ -267,26 +229,18 @@ function aliasTopLevelFields(prefix: string, doc: DocumentNode): DocumentNode {
  *     ... on Query { gatsby1_bar: bar }
  *   }
  */
-function aliasFieldsInSelection(
-  prefix: string,
-  selections: ReadonlyArray<SelectionNode>,
-  document: DocumentNode
-): SelectionNode[] {
-  return _.flatMap(selections, (selection: SelectionNode): SelectionNode[] => {
+function aliasFieldsInSelection(prefix, selections, document) {
+  return _.flatMap(selections, selection => {
     switch (selection.kind) {
       case Kind.INLINE_FRAGMENT:
         return [aliasFieldsInInlineFragment(prefix, selection, document)]
-
       case Kind.FRAGMENT_SPREAD: {
         const inlineFragment = inlineFragmentSpread(selection, document)
         return [
           addSkipDirective(selection),
           aliasFieldsInInlineFragment(prefix, inlineFragment, document),
-          // Keep original spread in selection with @skip(if: true)
-          // otherwise if this was a single fragment usage the query will fail validation
         ]
       }
-
       case Kind.FIELD:
       default:
         return [aliasField(selection, prefix)]
@@ -294,12 +248,8 @@ function aliasFieldsInSelection(
   })
 }
 
-interface INodeWithDirectives {
-  readonly directives?: ReadonlyArray<DirectiveNode>
-}
-
-function addSkipDirective<T extends INodeWithDirectives>(node: T): T {
-  const skipDirective: DirectiveNode = {
+function addSkipDirective(node) {
+  const skipDirective = {
     kind: Kind.DIRECTIVE,
     name: { kind: Kind.NAME, value: `skip` },
     arguments: [
@@ -325,11 +275,7 @@ function addSkipDirective<T extends INodeWithDirectives>(node: T): T {
  * To
  *   ... on Query { gatsby1_foo: foo, ... on Query { gatsby1_bar: foo } }
  */
-function aliasFieldsInInlineFragment(
-  prefix: string,
-  fragment: InlineFragmentNode,
-  document: DocumentNode
-): InlineFragmentNode {
+function aliasFieldsInInlineFragment(prefix, fragment, document) {
   const { selections } = fragment.selectionSet
   return {
     ...fragment,
@@ -350,10 +296,7 @@ function aliasFieldsInInlineFragment(
  * Transforms to:
  *   query { ... on Query { bar } }
  */
-function inlineFragmentSpread(
-  spread: FragmentSpreadNode,
-  document: DocumentNode
-): InlineFragmentNode {
+function inlineFragmentSpread(spread, document) {
   const fragment = document.definitions.find(
     def =>
       def.kind === Kind.FRAGMENT_DEFINITION &&
@@ -362,8 +305,7 @@ function inlineFragmentSpread(
   if (!fragment) {
     throw new Error(`Fragment ${spread.name.value} does not exist`)
   }
-  const { typeCondition, selectionSet } = fragment as FragmentDefinitionNode
-
+  const { typeCondition, selectionSet } = fragment
   return {
     kind: Kind.INLINE_FRAGMENT,
     typeCondition,
@@ -372,10 +314,7 @@ function inlineFragmentSpread(
   }
 }
 
-function prefixNodeName<T extends { name: NameNode }>(
-  namedNode: T,
-  prefix: string
-): T {
+function prefixNodeName(namedNode, prefix) {
   return {
     ...namedNode,
     name: {
@@ -392,7 +331,7 @@ function prefixNodeName<T extends { name: NameNode }>(
  *   { foo } -> { gatsby1_foo: foo }
  *   { foo: bar } -> { gatsby1_foo: bar }
  */
-function aliasField(field: FieldNode, aliasPrefix: string): FieldNode {
+function aliasField(field, aliasPrefix) {
   const aliasNode = field.alias ? field.alias : field.name
   return {
     ...field,
@@ -403,10 +342,10 @@ function aliasField(field: FieldNode, aliasPrefix: string): FieldNode {
   }
 }
 
-function isQueryDefinition(def): def is OperationDefinitionNode {
+function isQueryDefinition(def) {
   return def.kind === Kind.OPERATION_DEFINITION && def.operation === `query`
 }
 
-function isFragmentDefinition(def): def is FragmentDefinitionNode {
+function isFragmentDefinition(def) {
   return def.kind === Kind.FRAGMENT_DEFINITION
 }

